File: heroku-jvm/src/main/java/com/riseofcat/server/TickGame.java
package com.riseofcat.server;
import com.riseofcat.lib.DefaultValueMap;
import com.riseofcat.share.mass.ClientPayload;
import com.riseofcat.share.mass.GameConst;
import com.riseofcat.share.mass.TickActions;
import com.riseofcat.share.mass.InStateAction;
import com.riseofcat.share.mass.ServerPayload;
import com.riseofcat.share.ShareTodo;
import com.riseofcat.share.base.Tick;
import com.riseofcat.share.mass.BigAction;
import com.riseofcat.share.mass.NewCarAction;
import com.riseofcat.share.mass.PlayerAction;
import com.riseofcat.share.mass.PlayerId;
import com.riseofcat.share.mass.State;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.ConcurrentHashMap;

public class TickGame {
private final long startTime = System.currentTimeMillis();
private int previousActionsVersion = 0;
volatile private int tick = 0;//todo volatile redundant? //todo float
private State state = new State();
private DefaultValueMap<Tick, List<Action>> actions = new DefaultValueMap<>(new ConcurrentHashMap<>(), ArrayList::new);
private Map<PlayerId, Integer> mapPlayerVersion = new ConcurrentHashMap<>();
public TickGame(ConcreteRoomsServer.Room room) {
	room.onPlayerAdded.add(player -> {
		synchronized(TickGame.this) {
			int d = 1;
			actions.getExistsOrPutDefault(new Tick(tick + d)).add(new Action(++previousActionsVersion, new NewCarAction(player.getId()).toBig()));
			ServerPayload payload = createStablePayload();
			payload.setWelcome(new ServerPayload.Welcome(player.getId()));
			payload.setActions(new ArrayList<>());
			for(Map.Entry<Tick, List<Action>> entry : actions.getMap().entrySet()) {
				ArrayList<BigAction> temp = new ArrayList<>();
				for(Action a : entry.getValue()) temp.add(a.pa);
				payload.getActions().add(new TickActions(entry.getKey().getTick(), temp));
			}
			player.session.send(payload);
			mapPlayerVersion.put(player.getId(), previousActionsVersion);
		}
		for(RoomsDecorator<ClientPayload, ServerPayload>.Room.Player p : room.getPlayers()) if(!p.equals(player)) updatePlayer(p);//Говорим другим, что пришёл новый игрок
	});
	room.onMessage.add(message -> {
		synchronized(TickGame.this) {
			if(message.payload.getActions() != null) {
				for(ClientPayload.ClientAction a : message.payload.getActions()) {
					ServerPayload payload = new ServerPayload(tick);
					int delay = 0;
					if(a.getTick() < getStableTick().getTick()) {
						if(a.getTick() < getRemoveBeforeTick()) {
							payload.setCanceled(new HashSet<>());
							payload.getCanceled().add(a.getAid());
							message.player.session.send(payload);//todo move out of for
							continue;
						} else delay = getStableTick().getTick() - a.getTick();
					} else if(a.getTick() > getFutureTick()) {
						payload.setCanceled(new HashSet<>());
						payload.getCanceled().add(a.getAid());
						message.player.session.send(payload);//todo move out of for
						continue;
					}
					payload.setApply(new ArrayList<>());
					payload.getApply().add(new ServerPayload.AppliedActions(a.getAid(), delay));
					actions.getExistsOrPutDefault(new Tick(a.getTick() + delay)).add(new Action(++previousActionsVersion, new PlayerAction(message.player.getId(), a.getAction()).toBig()));
					if(ShareTodo.INSTANCE.getSIMPLIFY()) updatePlayerInPayload(payload, message.player);
					message.player.session.send(payload);//todo move out of for
				}
			}
		}
		for(RoomsDecorator<ClientPayload, ServerPayload>.Room.Player p : room.getPlayers()) if(!p.equals(message.player)) updatePlayer(p);
	});
	Timer timer = new Timer();
	timer.schedule(new TimerTask() {
		@Override
		public void run() {
			class Adapter implements Iterator<InStateAction> {
				private Iterator<Action> iterator;
				public Adapter(List<Action> arr) {
					if(arr != null) iterator = arr.iterator();
				}
				public boolean hasNext() {
					return iterator != null && iterator.hasNext();
				}
				public InStateAction next() {
					return iterator.next().pa;
				}
			}
			while(System.currentTimeMillis() - startTime > tick * GameConst.INSTANCE.getUPDATE_MS()) {
				synchronized(TickGame.this) {
					state.act(new Adapter(actions.getMap().get(getStableTick()))).tick();
					TickGame.this.actions.getMap().remove(getStableTick());
					++tick;
					if(tick % 200 == 0) /*todo %*/ for(ConcreteRoomsServer.Room.Player player : room.getPlayers()) player.session.send(createStablePayload());
				}
			}
		}
	}, 0, GameConst.INSTANCE.getUPDATE_MS() / 2);
}
private void updatePlayer(RoomsDecorator<ClientPayload, ServerPayload>.Room.Player p) {
	ServerPayload payload = new ServerPayload(tick);
	updatePlayerInPayload(payload, p);
	p.session.send(payload);
}
private void updatePlayerInPayload(ServerPayload payload, RoomsDecorator<ClientPayload, ServerPayload>.Room.Player p) {
	payload.setActions(new ArrayList<>());
	synchronized(this) {
		payload.setTick(tick);//todo redundant? but synchronized
		for(Map.Entry<Tick, List<Action>> entry : actions.getMap().entrySet()) {
			ArrayList<BigAction> temp = new ArrayList<>();
			for(Action a : entry.getValue()) if(ShareTodo.INSTANCE.getSIMPLIFY() || a.pa.getP() == null || !a.pa.getP().getId().equals(p.getId())) if(a.actionVersion > mapPlayerVersion.get(p.getId())) temp.add(a.pa);
			if(temp.size() > 0) payload.getActions().add(new TickActions(entry.getKey().getTick(), temp));
		}
		mapPlayerVersion.put(p.getId(), previousActionsVersion);
	}
}
ServerPayload createStablePayload() {
	ServerPayload result = new ServerPayload(tick);
	result.setStable(new ServerPayload.Stable(getStableTick().getTick(), state));
	return result;
}
private Tick getStableTick() {
	int result = tick - GameConst.INSTANCE.getDELAY_TICKS() + 1;
	if(result < 0) return new Tick(0);
	return new Tick(result);
}
private int getRemoveBeforeTick() {
	return tick - GameConst.INSTANCE.getREMOVE_TICKS() + 1;
}
private int getFutureTick() {
	return tick + GameConst.INSTANCE.getFUTURE_TICKS();
}
private static class ConcreteRoomsServer extends RoomsDecorator<ClientPayload, ServerPayload> {

}
private class Action {
	public int actionVersion;
	public BigAction pa;
	public Action(int actionVersion, BigAction pa) {
		this.actionVersion = actionVersion;
		this.pa = pa;
	}
}
private void todo() {
	ConcreteRoomsServer.Room.Player player = null;
	long startTime = player.session.get(UsageMonitorDecorator.Extra.class).getStartTime();
	Integer latency = player.session.get(PingDecorator.Extra.class).getLatency();
}
}

File: heroku-jvm/src/main/java/com/riseofcat/server/TickGame.kt
package com.riseofcat.server

import com.riseofcat.client.*
import com.riseofcat.common.*
import com.riseofcat.lib.*
import com.riseofcat.share.mass.*

class TickGame(val room:RoomsDecorator<ClientPayload,ServerPayload>.Room) {
  val STABLE_STATE_UPDATE:Duration? = Duration(10_000)
  @Volatile private var previousActionsVersion = 0//todo почитать про volatile в Kotlin
  private val state = State()
  private var actions:MutableList<Action> = Common.createConcurrentList()
  private val mapPlayerVersion = Common.createConcurrentHashMap<PlayerId,Int>()
  private var previousStableUpdate:TimeStamp = lib.time

  val realtimeTick get() = Tick((lib.time - room.createTime)/GameConst.UPDATE)
  //todo averageLatency считать умнее
  val averageLatency get() = room.getPlayers().sumByDuration{it.session.get(PingDecorator.Extra::class.java)!!.lastPingDelay?: DEFAULT_LATENCY}/room.getPlayers().size
  val recommendedLatency get() = averageLatency*2
  val recommendedDelay get() = Tick(recommendedLatency/GameConst.UPDATE + 1)
  val maxDelay get() = recommendedDelay*2
  val newCarDelay get() = recommendedDelay*2
  val removeAfterDelay get() = recommendedDelay*3//0 - значит всё что позже stable - удаляется

  init {
    room.onPlayerAdded.add {player->
      updateGame()
      redundantSynchronize(this@TickGame) {
        actions.add(Action(++previousActionsVersion,TickAction(realtimeTick+newCarDelay,player.id,n = NewCarAction(player.id))))
        val payload = createStablePayload(Welcome(player.id, room.createTime))
        player.session.send(payload)
      }
      for(p in room.getPlayers()) updatePlayer(p)//Говорим другим, что пришёл новый игрок
    }
    room.onMessage.add {message->
      updateGame()
      redundantSynchronize(this@TickGame) {
        for(a in message.payload.actions) {
          val payload = ServerPayload(recommendedLatency = recommendedLatency)
          var delay = Tick(0)
          if(a.tick<state.tick) {
            if(a.tick <= state.tick-removeAfterDelay) continue
            delay = state.tick-a.tick//задерживаем
          }
          actions.add(Action(++previousActionsVersion,TickAction(a.tick+delay, message.player.id, p = PlayerAction(message.player.id,a.action))))
          updatePlayerInPayload(payload,message.player)
          message.player.session.send(payload)//todo move out of for
        }
      }
      for(p in room.getPlayers()) if(p!=message.player) updatePlayer(p)
    }
  }

  private fun updateGame() {
    fun condition() = state.tick<realtimeTick-maxDelay
    if(!condition()) return

    val test = true//todo test performance
    synchronized(this@TickGame) {
      while(condition()) {
        state act actions.map {it.ta}.filter {it.tick==state.tick}.iterator()
        if(!test)actions.removeAll {it.ta.tick==state.tick}
        state.tick()
      }
      if(test)actions.removeAll {it.ta.tick<state.tick}

      if(STABLE_STATE_UPDATE!=null&&lib.time>previousStableUpdate+STABLE_STATE_UPDATE) {
        previousStableUpdate = lib.time
        for(player in room.getPlayers()) player.session.send(createStablePayload())
      }
    }
  }

  private fun updatePlayer(p:RoomsDecorator<ClientPayload,ServerPayload>.Room.Player) {
    val payload = ServerPayload()
    updatePlayerInPayload(payload,p)
    p.session.send(payload)
  }

  private fun updatePlayerInPayload(payload:ServerPayload,p:RoomsDecorator<ClientPayload,ServerPayload>.Room.Player) {
    redundantSynchronize(this) {
      val filtered = actions.filter {it.actionVersion>mapPlayerVersion[p.id] ?: 0}
      payload.actions = filtered.map {it.ta}
      val maxActionVer = filtered.map {it.actionVersion}.max()
      if(maxActionVer != null) mapPlayerVersion.put(p.id,maxActionVer)
    }
  }

  internal fun createStablePayload(welcome:Welcome?=null):ServerPayload = ServerPayload(
    welcome =  welcome,
    stable = state,
    recommendedLatency = recommendedLatency
  )
}

inline fun <R> redundantSynchronize(lock:Any,block:()->R) = if(true) block() else synchronized(lock,block) //todo test performance
private class Action(val actionVersion:Int, val ta:TickAction)

